src/util/buf.rs

```diff
@@ -229,7 +229,11 @@ impl PipeBuffer {
                 }
             }
         }
-        self.end = end;
+        if end == self.buf.cap() {
+            self.end = 0;
```

Comment:
This can't happen. If `start = 0`, and `end = self.buf.cap()`, then the buffer will contain `capacity + 1` data, which is guaranteed not happening by slice_append.